File: app/src/main/java/com/rogger/bp/ui/add/AddFragment.java
package com.rogger.bp.ui.add;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.bumptech.glide.Glide;
import com.rogger.bp.R;
import com.rogger.bp.data.database.FirebaseStorageDataSource;
import com.rogger.bp.data.database.ProdutoImagemDataSource;
import com.rogger.bp.data.model.Categoria;
import com.rogger.bp.data.model.Produto;
import com.rogger.bp.data.model.ProdutoImagem;
import com.rogger.bp.data.repository.ProdutoImagemRepository;
import com.rogger.bp.databinding.FragmentAddBinding;
import com.rogger.bp.ui.base.DialogUtil;
import com.rogger.bp.ui.base.Utils;
import com.rogger.bp.ui.commun.ShowSelectDialog;
import com.rogger.bp.ui.gallery.CameraCallback;
import com.rogger.bp.ui.gallery.ImagePikerUtil;
import com.rogger.bp.ui.gallery.ImageUtils;
import com.rogger.bp.ui.scanner.ImageBarcode;
import com.rogger.bp.ui.viewmodel.CategoriaViewModel;
import com.rogger.bp.ui.viewmodel.DataViewModel;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class AddFragment extends Fragment {
    private FragmentAddBinding binding;
    private long timestamp;
    private Produto produto;
    private File photoFile;
    private boolean confirmed = false;
    private boolean carregandoSpinner = true;
    private int categoriaId = -1;
    private String imagemUrlGlobal = null;
    private ArrayAdapter<Categoria> categoriaAdapter;
    private List<Categoria> listaCategorias = new ArrayList<>();
    private ImagePikerUtil cameraUtil;
    private ActivityResultLauncher<Intent> cameraLauncher;
    private CategoriaViewModel categoriaViewModel;
    private DataViewModel dataViewModel;
    private ProdutoImagemRepository produtoImagemRepository;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentAddBinding.inflate(inflater, container, false);

        categoriaViewModel = new ViewModelProvider(requireActivity()).get(CategoriaViewModel.class);
        dataViewModel = new ViewModelProvider(requireActivity()).get(DataViewModel.class);
        produtoImagemRepository = ProdutoImagemRepository.getInstance();
        produto = new Produto();
        cameraUtil = new ImagePikerUtil();

        String dataAtual = Utils.getCurrentDateFormatted();
        binding.datePickerBtnAdd.setText(dataAtual);
        binding.txtAddData.setText(dataAtual);
        timestamp = Utils.getCurrentTimestamp();

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        configurarSpinner();
        configurarGalleryResult();
        configurarCamera();
        configurarListeners();
        configurarMenu();

        // Recebe o código de barras vindo do scanner
        Bundle args = getArguments();
        if (args != null) {
            String barcode = args.getString("key_barcode");
            categoriaId = args.getInt("categoria_id", -1);

            if (barcode != null && !barcode.isEmpty()) {
                binding.txtAddBarcode.setText(barcode);
                // ─── NOVO FLUXO: consulta imagem global antes de exibir formulário
                consultarImagemGlobal(barcode);
            }
        }
    }

    private void consultarImagemGlobal(@NonNull String barcode) {
        mostrarProgressoConsulta(true);

        produtoImagemRepository.buscarPorCodigoBarras(barcode,
                new ProdutoImagemDataSource.BuscaCallback() {

                    @Override
                    public void onEncontrado(@NonNull ProdutoImagem produtoImagem) {
                        if (!isAdded()) return;
                        requireActivity().runOnUiThread(() -> {
                            mostrarProgressoConsulta(false);
                            preencherComImagemExistente(produtoImagem);
                        });
                    }

                    @Override
                    public void onNaoEncontrado() {
                        if (!isAdded()) return;
                        requireActivity().runOnUiThread(() -> {
                            mostrarProgressoConsulta(false);
                        });
                    }

                    @Override
                    public void onErro(@NonNull Exception e) {
                        if (!isAdded()) return;
                        requireActivity().runOnUiThread(() -> {
                            mostrarProgressoConsulta(false);
                        });
                    }
                });
    }

    private void preencherComImagemExistente(@NonNull ProdutoImagem produtoImagem) {
        imagemUrlGlobal = produtoImagem.getImagemUrl();
        if (produtoImagem.temImagem()) {
            Glide.with(this)
                    .load(imagemUrlGlobal)
                    .override(500, 500)
                    .placeholder(R.drawable.carregando)
                    .error(R.drawable.imagem_error)
                    .centerCrop()
                    .into(binding.fragmentImgAdd);
            binding.fragmentImgAdd.setClickable(false);
        }
        if (produtoImagem.temNome()) {
            binding.edtNameFragmentAdd.setText(produtoImagem.getNomeProduto());
            binding.edtNameFragmentAdd.setSelection(
                    binding.edtNameFragmentAdd.getText().length());
        }
    }

    private void saveData() {
        produto.setNome(binding.edtNameFragmentAdd.getText().toString().trim());
        produto.setAnotacoes(binding.editFragmentNoteAdd.getText().toString());
        produto.setTimestamp(timestamp);
        produto.setCodigoBarras(binding.txtAddBarcode.getText().toString().trim());
        if (imagemUrlGlobal != null && !imagemUrlGlobal.isEmpty()) {
            produto.setImagem(imagemUrlGlobal);
            dataViewModel.insert(produto, null);
            return;
        }

        if (photoFile != null && !produto.getCodigoBarras().isEmpty()) {
            produto.setImagem(photoFile.getAbsolutePath());
            mostrarProgressoUpload(true);

            dataViewModel.insert(produto, new FirebaseStorageDataSource.UploadCallback() {

                @Override
                public void onProgresso(int porcentagem) {
                    if (!isAdded()) return;
                    requireActivity().runOnUiThread(() ->
                            binding.progressUploadAdd.setProgress(porcentagem));
                }

                @Override
                public void onSucesso(String urlDownload) {
                    if (!isAdded()) return;
                    requireActivity().runOnUiThread(() -> {
                        mostrarProgressoUpload(false);
                    });
                }

                @Override
                public void onErro(Exception e) {
                    if (!isAdded()) return;
                    requireActivity().runOnUiThread(() -> {
                        mostrarProgressoUpload(false);
                    });
                }
            });
            return;
        }
        dataViewModel.insert(produto, null);
    }

    private void configurarSpinner() {
        categoriaViewModel.getCategories().observe(getViewLifecycleOwner(), categories -> {
            listaCategorias.clear();
            if (categories != null) listaCategorias.addAll(categories);

            categoriaAdapter = new ArrayAdapter<>(
                    requireContext(),
                    android.R.layout.simple_spinner_item,
                    listaCategorias);
            categoriaAdapter.setDropDownViewResource(
                    android.R.layout.simple_spinner_dropdown_item);
            binding.spinnerAdd.setAdapter(categoriaAdapter);

            if (categoriaId != -1) selecionarCategoriaPorId(categoriaId);
            carregandoSpinner = false;
        });

        binding.spinnerAdd.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view,
                                       int position, long id) {
                if (carregandoSpinner) return;
                produto.setCategoryId(listaCategorias.get(position).getId());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
    }

    private void configurarGalleryResult() {
        getParentFragmentManager().setFragmentResultListener(
                "gallery_result",
                getViewLifecycleOwner(),
                (key, bundle) -> {
                    try {
                        Uri img = Uri.parse(bundle.getString("imageUri"));
                        File file = ImagePikerUtil.createImageFile(requireContext());
                        photoFile = ImageUtils.processImage(requireContext(), img, file);
                        // Imagem local selecionada — descarta qualquer URL global anterior
                        imagemUrlGlobal = null;
                        Glide.with(this).load(photoFile).override(500, 500).centerCrop().into(binding.fragmentImgAdd);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
    }

    private void configurarCamera() {
        cameraLauncher = cameraUtil.register(this, new CameraCallback() {
            @Override
            public void onImageCaptured(@NonNull Uri imageUri, @NonNull File imageFile) {
                try {
                    photoFile = ImageUtils.processImage(requireContext(), imageUri, imageFile);
                    imagemUrlGlobal = null;
                    Glide.with(this).load(photoFile).override(500, 500).centerCrop().into(binding.fragmentImgAdd);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public void onCancel() {
            }

            @Override
            public void onError(@NonNull Exception e) {
                Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_LONG).show();
            }
        });
    }

    private void configurarListeners() {
        binding.datePickerBtnAdd.setOnClickListener(v ->
                Utils.showDatePicker(requireContext(), (ts, dataFormatada) -> {
                    binding.datePickerBtnAdd.setText(dataFormatada);
                    binding.txtAddData.setText(dataFormatada);
                    timestamp = ts;
                }));

        binding.txtAddBarcode.setOnClickListener(v ->
                startImageBarcode(binding.txtAddBarcode.getText().toString()));

        binding.fragmentImgAdd.setOnClickListener(v -> {
            if (imagemUrlGlobal != null) return;
            confirmed = false;
            ShowSelectDialog.show(getContext(), new ShowSelectDialog.selectedCallback() {
                @Override
                public void openGallery() {
                    Navigation.findNavController(requireView())
                            .navigate(R.id.action_addFragment_to_nav_gallery_fragment);
                }

                @Override
                public void openCamera() {
                    if (photoFile != null) photoFile.delete();
                    cameraUtil.openCamera(requireContext(), cameraLauncher);
                }
            });
        });

        binding.btnFgmSaveAdd.setOnClickListener(v -> {
            if (Utils.validEditText(binding.edtNameFragmentAdd)) {
                saveData();
                confirmed = true;
                Toast.makeText(getContext(), "Produto salvo com sucesso", Toast.LENGTH_SHORT).show();
                Navigation.findNavController(requireView()).popBackStack();
            }
        });
    }

    private void configurarMenu() {
        requireActivity().addMenuProvider(new MenuProvider() {
            @Override
            public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
                menu.clear();
                menuInflater.inflate(R.menu.menu_edit, menu);
            }

            @Override
            public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
                if (menuItem.getItemId() == R.id.action_delete) {
                    DialogUtil.showDeleteDialog(
                            requireContext(),
                            "Deseja realmente excluir este produto?",
                            () -> {
                                ImagePikerUtil.cleanUpTempFiles(photoFile);
                                Navigation.findNavController(requireView()).popBackStack();
                            });
                    return true;
                }
                return false;
            }
        }, getViewLifecycleOwner(), Lifecycle.State.RESUMED);
    }

    private void mostrarProgressoConsulta(boolean mostrar) {
        binding.progressUploadAdd.setVisibility(mostrar ? View.VISIBLE : View.GONE);
        binding.btnFgmSaveAdd.setEnabled(!mostrar);
    }

    private void mostrarProgressoUpload(boolean mostrar) {
        binding.progressUploadAdd.setVisibility(mostrar ? View.VISIBLE : View.GONE);
        binding.progressUploadAdd.setIndeterminate(!mostrar);
        binding.btnFgmSaveAdd.setEnabled(!mostrar);
    }

    private void startImageBarcode(String barcode) {
        Intent intent = new Intent(getContext(), ImageBarcode.class);
        intent.putExtra("keyBarcode", barcode);
        startActivity(intent);
    }

    private void selecionarCategoriaPorId(int id) {
        for (int i = 0; i < listaCategorias.size(); i++) {
            if (listaCategorias.get(i).getId() == id) {
                binding.spinnerAdd.setSelection(i);
                produto.setCategoryId(id);
                return;
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (!confirmed && photoFile != null && imagemUrlGlobal == null) {
            ImagePikerUtil.cleanUpTempFiles(photoFile);
        }
    }
}

File: app/src/main/java/com/rogger/bp/ui/add/view/Add_Fragment.kt
package com.rogger.bp.ui.add.view

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.bumptech.glide.Glide
import com.rogger.bp.R
import com.rogger.bp.data.model.PostCategory
import com.rogger.bp.data.model.PostImage
import com.rogger.bp.data.model.PostProduct
import com.rogger.bp.data.model.UserAuth
import com.rogger.bp.databinding.FragmentAddBinding
import com.rogger.bp.ui.add.RegisterAdd
import com.rogger.bp.ui.add.presentation.AddItemPresenter
import com.rogger.bp.ui.animation.ToastCustom
import com.rogger.bp.ui.base.Utils
import com.rogger.bp.ui.commun.DependencyInjector
import com.rogger.bp.ui.commun.ShowSelectDialog
import com.rogger.bp.ui.gallery.CameraCallback
import com.rogger.bp.ui.gallery.ImagePikerUtil
import com.rogger.bp.ui.gallery.ImageUtils
import com.rogger.bp.ui.scanner.ImageBarcode
import java.io.File

class AddItemFragment : Fragment(R.layout.fragment_add),RegisterAdd.View {
    companion object {
        const val KEY_BARCODE     = "key_barcode"
        const val KEY_CATEGORIA   = "categoria_id"
    }
    // ── MVP ───────────────────────────────────────────────────────────────
    override lateinit var presenter: RegisterAdd.Presenter

    private var _binding: FragmentAddBinding? = null
    private val binding get() = _binding!!
    private var listaCategorias: List<PostCategory> = emptyList()
    private var spinnerPronto = false

    private var barcode: String = ""
    private var timestamp: Long = 0L
    private var photoFile: File? = null
    private var remoteImageUri: String? = null   // URI da imagem já existente no Storage
    private var confirmed: Boolean = false        // evita apagar foto se o usuário salvou

    private val cameraUtil = ImagePikerUtil()
    private lateinit var cameraLauncher: ActivityResultLauncher<Intent>

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentAddBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val repository = DependencyInjector.registerProductRepository()
        val categoryRepository = DependencyInjector.registerCategoryRepository(requireContext())
        presenter = AddItemPresenter(this, repository, categoryRepository)

        val dataAtual = Utils.getCurrentDateFormatted()
        binding.datePickerBtnAdd.text = dataAtual
        binding.txtAddData.text = dataAtual
        timestamp = Utils.getCurrentTimestamp()

        setupCamera()
        setupGalleryResult()
        setupListeners()
        readArguments()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        // Limpa arquivo temporário se o usuário cancelou sem salvar
        if (!confirmed && photoFile != null && remoteImageUri == null) {
            ImagePikerUtil.cleanUpTempFiles(photoFile)
        }
        _binding = null
    }

    override fun onDestroy() {
        if (::presenter.isInitialized) presenter.onDestroy()
        super.onDestroy()
    }
    private fun setupCamera() {
        cameraLauncher = cameraUtil.register(this, object : CameraCallback {
            override fun onImageCaptured(imageUri: Uri, imageFile: File) {
                try {
                    photoFile = ImageUtils.processImage(requireContext(), imageUri, imageFile)
                    remoteImageUri = null    // foto local substitui qualquer URI remota
                    showLocalImage(photoFile!!)
                    // Upload é feito apenas ao clicar em "Salvar" (saveProduct).
                } catch (e: Exception) {
                    onFailure(e.message ?: "Erro ao processar imagem")
                }
            }

            override fun onCancel() { /* nenhuma ação necessária */ }
            override fun onError(e: Exception) {
                onFailure(e.message ?: "Erro na câmera")
            }
        })
    }

    private fun setupGalleryResult() {
        parentFragmentManager.setFragmentResultListener(
            "gallery_result",
            viewLifecycleOwner
        ) { _, bundle ->
            try {
                val img = Uri.parse(bundle.getString("imageUri"))
                val file = ImagePikerUtil.createImageFile(requireContext())
                photoFile = ImageUtils.processImage(requireContext(), img, file)
                remoteImageUri = null
                showLocalImage(photoFile!!)
                // Upload é feito apenas ao clicar em "Salvar" (saveProduct),
                // nunca antes de o produto existir no banco.
            } catch (e: Exception) {
                onFailure(e.message ?: "Erro ao processar imagem da galeria")
            }
        }
    }

    private fun setupListeners() {
        binding.btnFgmSaveAdd.setOnClickListener {
            saveProduct()
        }
        binding.datePickerBtnAdd.setOnClickListener {
            Utils.showDatePicker(requireContext()) { ts, dataFormatada ->
                binding.datePickerBtnAdd.text = dataFormatada
                binding.txtAddData.text = dataFormatada
                timestamp = ts
            }
        }

        binding.txtAddBarcode.setOnClickListener {
            openBarcodeImageScreen(binding.txtAddBarcode.text.toString())
        }

        binding.fragmentImgAdd.setOnClickListener {
            if (remoteImageUri != null) return@setOnClickListener   // bloqueado
            openMediaPicker()
        }
    }

    private fun readArguments() {
        val args = arguments ?: return

        barcode = args.getString(KEY_BARCODE).orEmpty()

        if (barcode.isNotEmpty()) {
            binding.txtAddBarcode.text = barcode
            // Verifica se já existe imagem para este barcode no catálogo global
            presenter.checkOrCreateImage(barcode)
        }
    }

    /** Exibe arquivo local no ImageView. */
    private fun showLocalImage(file: File) {
        Glide.with(requireContext())
            .load(file)
            .override(500, 500)
            .centerCrop()
            .into(binding.fragmentImgAdd)
    }

    /** Exibe URL remota no ImageView. */
    private fun showRemoteImage(url: String) {
        Glide.with(requireContext())
            .load(url)
            .override(500, 500)
            .placeholder(R.drawable.carregando)
            .error(R.drawable.imagem_error)
            .centerCrop()
            .into(binding.fragmentImgAdd)
        binding.fragmentImgAdd.isClickable = false
    }

    private fun openMediaPicker() {
        ShowSelectDialog.show(requireContext(), object : ShowSelectDialog.selectedCallback {
            override fun openGallery() {
                this@AddItemFragment.openGallery()
            }
            override fun openCamera() {
                this@AddItemFragment.openCamera()
            }
        })
    }

    private fun saveProduct() {
        val nome = binding.edtNameFragmentAdd.text.toString().trim()
        if (!Utils.validEditText(binding.edtNameFragmentAdd)) return

        val imageUri: Uri = when {
            remoteImageUri != null -> Uri.parse(remoteImageUri)
            photoFile != null      -> Uri.fromFile(photoFile)
            else                   -> Uri.EMPTY
        }

        val product = PostProduct(
            imageUri   = imageUri.toString(),
            name       = nome,
            note       = binding.editFragmentNoteAdd.text.toString(),
            barcode    = barcode,
            deleted    = false,
            timestamp  = timestamp,
            publisher  = UserAuth("", "", "", "", null)  // preenchido no DataSource com Auth atual
        )

        confirmed = true
        presenter.saveProduct(product)
    }

    private fun openBarcodeImageScreen(barcodeValue: String) {
        val intent = Intent(requireContext(), ImageBarcode::class.java)
        intent.putExtra("keyBarcode", barcodeValue)
        startActivity(intent)
    }
    override fun showProgress(enable: Boolean) {
        binding.progressUploadAdd.visibility = if (enable) View.VISIBLE else View.GONE
        binding.btnFgmSaveAdd.isEnabled = !enable
    }
    override fun onFailure(message: String) {
        Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT).show()
    }
    override fun imageAlreadyExists(postImage: PostImage) {
        remoteImageUri = postImage.uri
        showRemoteImage(postImage.uri)
        if (postImage.name.isNotEmpty()) {
            binding.edtNameFragmentAdd.setText(postImage.name)
            binding.edtNameFragmentAdd.setSelection(
                binding.edtNameFragmentAdd.text.length
            )
        }
    }
    override fun onImageNotFound() {
        binding.fragmentImgAdd.isClickable = true
    }
    override fun openCamera() {
        photoFile?.delete()
        cameraUtil.openCamera(requireContext(), cameraLauncher)
    }
    override fun openGallery() {
        findNavController().navigate(R.id.action_addFragment_to_nav_gallery_fragment)
    }
    override fun goToHome() {
        ToastCustom.showCustomToast(requireContext(),"")

        findNavController().popBackStack()
    }

    override fun showCategories(categories: List<PostCategory>) {
        listaCategorias = categories

        val nomes = mutableListOf("Selecione uma categoria")
        nomes.addAll(categories.map { it.name })

        binding.spinnerAdd.adapter = ArrayAdapter(
            requireContext(),
            android.R.layout.simple_spinner_item,
            nomes
        ).also { it.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item) }

        spinnerPronto = true

        listaCategorias.forEachIndexed { index, cat ->
            if (cat.firestoreId.isBlank()) {
                binding.spinnerAdd.setSelection(index + 1)
                return
            }
        }
    }
}